test/boost/row_cache_test.cc

```diff
@@ -4814,3 +4814,115 @@ SEASTAR_THREAD_TEST_CASE(test_preempt_cache_update) {
+    mutation m(s, pk);
+    m.set_clustered_cell(ck1, "v", data_value(0), 1);
+    m.partition().apply_delete(*s, ck2, tombstone(1, dt_exp));
+    m.set_clustered_cell(ck3, "v", data_value(0), 1);
+
+    memtable_snapshot_source underlying(s);
+    underlying.apply(m);
+
+    cache_tracker tracker;
+    row_cache cache(s, snapshot_source([&] { return underlying(); }), tracker);
+    cache.populate(m);
+
+    with_allocator(tracker.allocator(), [&] {
+        auto& e = *cache.lookup(pk).partition().version()->partition().clustered_rows().begin();
+        tracker.get_lru().remove(e);
+        e.on_evicted(tracker);
+    });
+
+    // We have set up the desired state.
+    // Now we do a reverse query over the partition.
+    // This query will remove the expired entry at ck2, leaving cursor's _latest_it dangling.
+    // Then, it will populate the cache with ck3.
+    // Before the fix for issue #18045, this caused a (ASAN-triggering) use-after-free,
+    // because _latest_it was deferenced during the population.
+
+    tombstone_gc_state gc_state(nullptr);
+    auto slice = make_legacy_reversed(s, s->full_slice());
+    auto rd = cache.make_reader(
+        s->make_reversed(),
+        semaphore.make_permit(),
+        dht::partition_range::make_singular(pk),
+        slice,
+        nullptr,
+        streamed_mutation::forwarding::no,
+        mutation_reader::forwarding::no,
+        &gc_state);
+    auto close_rd = deferred_close(rd);
+    read_mutation_from_flat_mutation_reader(rd).get();
+}
+
+    auto dt_exp = gc_clock::now() - std::chrono::seconds(s->gc_grace_seconds().count() + 1);
+
+    cache_tracker tracker;
+
+    mutation m1(s, pk);
+    m1.set_clustered_cell(ck1, "v", data_value(0), 1);
+    m1.partition().apply_delete(*s, ck2, tombstone(1, dt_exp));
+    m1.set_clustered_cell(ck3, "v", data_value(0), 1);
+    fmt::print("{}", mutation_partition::printer(*s, m1.partition()));
+
+    memtable_snapshot_source underlying(s);
+    underlying.apply(m1);
+
+    row_cache cache(s, snapshot_source([&] { return underlying(); }), tracker);
+    cache.populate(m1);
+    with_allocator(tracker.allocator(), [&] {
+        auto& e = *cache.lookup(pk).partition().version()->partition().clustered_rows().begin();
+        tracker.get_lru().remove(e);
+        e.on_evicted(tracker);
+    });
+
+    tombstone_gc_state gc_state(nullptr);
+    auto slice = make_legacy_reversed(s, s->full_slice());
+    auto rd = cache.make_reader(
+        s->make_reversed(),
+        semaphore.make_permit(),
+        dht::partition_range::make_singular(pk),
+        slice,
+        nullptr,
+        streamed_mutation::forwarding::no,
+        mutation_reader::forwarding::no,
+        &gc_state);
+    auto close_rd = deferred_close(rd);
+
+    {
+        cache_entry& entry = cache.lookup(pk);
+        auto& cp = entry.partition().version()->partition();
+        fmt::print("{}", mutation_partition_v2::printer(*s, cp));
```

Comment:
Why not testlog.info()? This way output will be aligned with other logs.

Reply:
Whoops, that's a leftover from debugging, I didn't mean to commit it.

Reply:
I made a git mistake. Fixed.